openedx_tagging/core/tagging/api.py

```diff
@@ -0,0 +1,102 @@
+""""
+Tagging API
+
+Anyone using the openedx_tagging app should use these APIs instead of creating
+or modifying the models directly, since there might be other related model
+changes that you may not know about.
+
+Please look at the models.py file for more information about the kinds of data
+are stored in this app.
+"""
+from typing import List, Type
+
+from django.db.models import QuerySet
+from django.utils.translation import gettext_lazy as _
+
+from .models import ObjectTag, Tag, Taxonomy
+
+
+def create_taxonomy(
+    name,
+    description=None,
+    enabled=True,
+    required=False,
+    allow_multiple=False,
+    allow_free_text=False,
+) -> Taxonomy:
+    """
+    Creates, saves, and returns a new Taxonomy with the given attributes.
+    """
+    return Taxonomy.objects.create(
+        name=name,
+        description=description,
+        enabled=enabled,
+        required=required,
+        allow_multiple=allow_multiple,
+        allow_free_text=allow_free_text,
+    )
+
+
+def get_taxonomies(enabled=True) -> QuerySet:
+    """
+    Returns a queryset containing the enabled taxonomies, sorted by name.
+    If you want the disabled taxonomies, pass enabled=False.
+    If you want all taxonomies (both enabled and disabled), pass enabled=None.
+    """
+    queryset = Taxonomy.objects.order_by("name", "id")
+    if enabled is None:
+        return queryset.all()
+    return queryset.filter(enabled=enabled)
+
+
+def get_tags(taxonomy: Taxonomy) -> List[Tag]:
+    """
+    Returns a list of predefined tags for the given taxonomy.
+
+    Note that if the taxonomy allows free-text tags, then the returned list will be empty.
+    """
+    return taxonomy.get_tags()
+
+
+def resync_tags() -> int:
+    """
+    Reconciles all stored ObjectTag entries with any changes made to their associated taxonomies and tags.
+    """
+    num_changed = 0
+    for object_tag in ObjectTag.objects.all():
```

Comment:
If there are a _lot_ of `ObjecTag` rows, this has the downside of loading them all from the database and then converting them all into python objects before iterating over them. It seems like using `for object_tag in ObjectTag.objects.all().iterator()` [is better](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#iterator):

> For a `QuerySet` which returns a large number of objects that you only need to access once, this can result in better performance and a significant reduction in memory.

However, it's still not as good as I hoped:

> MySQL doesn’t support streaming results, hence the Python database driver loads the entire result set into memory. 

What?! Yet another point for postgres...

Reply:
@bradenmacdonald Cool, I didn't know about that! Will fix this.

>> MySQL doesn’t support streaming results, hence the Python database driver loads the entire result set into memory.
> 
> What?! Yet another point for postgres...

My favourite related comment so far came from this issue: https://code.djangoproject.com/ticket/2495#comment:14

>  After eight years, and considering that the fix is far from easy, could we file this under "use a better database" and close the ticket?